enhancements/cloud-integration/tokenized-auth-enablement-operators-on-cloud.md

```diff
@@ -0,0 +1,340 @@
+
+Operator installation under the proposed system:
+1. user goes to console and starts operator install
+2. console detects token-based (STS) cluster, and token-based-supporting operator
+3. console prompts user to create roles+supply ARN-type string
+4. console creates subscription with ARN-type string embedded as a spec.config.env
+5. operator deployment is created with ARN-type string in env
+6. operator creates `CredentialsRequest` including the ARN-type string
+7. Cloud Credential Operator populates Secret based on `CredentialsRequest`
+8. operator loads Secret, makes cloud service requests
+
+
+For the operator author team:
+- Add CredentialRequests to the bundle, known location;
+- Add "role-arn"-type fields to the CredentialRequests
+- Add references to the CredentialRequests in the CRD spec;
+- Use SDK to validate bundle to catch permission changes or mis-configuration
+- Add eventing to report status on a CR to indicate lacking STS credentials for fully operational deploy or update.
+
+For the Cloud Credential Operator:
+- Add a Token mode that will watch for CredentialRequests (as per mint mode) and then resolve time-based tokens
+  per cloud platform. 
+- Make Secret resource available for the operator via pod identity webhook. 
+- Pod identity webhook logic to account for per-operand tenancy permissions.
+
+For Operator Administrator:
+- Supply cloud credentials for STS (ARN ID, etc) in the known location and change CR to reflect this.
+
+For OperatorHub, Console:
+- Prompt for missing credentials that prevent fully operational install on STS cluster. Same for upgrades.
+- Badge indicating STS enabled operators (including operators that don't need cloud credentials)
+
+This graph shows the system as it needs to work (AWS only), not the proposed solution workflow:
+```mermaid
+graph  LR
+  sa([ServiceAccount Signing Keys]) -- Public <br> Key -->s3[S3 Bucket w/ OIDC Config];
+  sa([ServiceAccount Signing Keys]) -- Private <br> Key -->token_signing;
+  service<-.->config
+  service<-.->Pod
+  Pod-->role
+  subgraph cluster [AWS]
+  s3;
+  iam_id[IAM Identity Provider] -- Trust Tokens<br>Signed -->s3;
+  role[IAM Role For Pod]<-->service[AWS Services];
+  end
+  subgraph cluster2 [OpenShift]
+  token_signing[Token Signing]-->projected[Projected<br>ServiceAccount<br>Token];
+  subgraph Pod
+  config[AWS Config File]-->projected
+  end
+  end
+  classDef plain fill:#ddd,stroke:#fff,stroke-width:4px,color:#000;
+  classDef k8s fill:#326ce5,stroke:#fff,stroke-width:4px,color:#fff;
+  classDef cluster fill:#fff,stroke:#bbb,stroke-width:2px,color:#326ce5;
+  classDef cluster2 fill:#fff,stroke:#bbb,stroke-width:2px,color:#326ce5;
+  class config,role,service,s3,projected,token_signing,iam_id,pod1,pod2 k8s;
+  class sa plain;
+  class cluster cluster;
+  class cluster2 cluster1;
+```
+
+### API Extensions
+
+Added to the CredentialsRequest API 
+
+Adding to the `spec` field, the following subfields:
+
+`cloudTokenString`;<br>
+`cloudTokenPath`
```

Comment:
Could you convert this to golang so that we can see and discuss what the fields would actually look like?

Reply:
Adding to this that different platforms may have different requirements than a singular `cloudTokenString` which represents the AWS IAM Role ARN with respect to AWS STS and I think we should embed fields within platform specific structs if possible.

For instance, a secret created for GCP (ref: [CCO doc](https://github.com/openshift/cloud-credential-operator/blob/master/docs/gcp_workload_identity.md#changes-in-the-credentials-secret-with-workload-identity)) will need to know the `audience` which appears to include the GCP project as well as the workload identity pool infrastructure:

```json
{
   "type": "external_account",
   // "audience" includes GCP project, workload identity pool / provider
   "audience": "//iam.googleapis.com/projects/123456789/locations/global/workloadIdentityPools/test-pool/providers/test-provider",
   "subject_token_type": "urn:ietf:params:oauth:token-type:jwt",
   "token_url": "https://sts.googleapis.com/v1/token",
   // "service_account_impersonation_url" includes specific service account that will be impersonated, the AWS Role ARN equivalent
   "service_account_impersonation_url": "https://iamcredentials.googleapis.com/v1/projects/-/serviceAccounts/test-service-account-42ssv@test-project.iam.gserviceaccount.com:generateAccessToken",
   // "credential_source" includes the path to the token as mounted on the deployment
   "credential_source": {
      "file": "/path/to/oidc/token",
      "format": {
         "type": "text"
      }
   }
}
```

Reply:
Please also re-add @JoelSpeed as API reviewer to EP. The `CredentialsRequest` should very arguably have been born into the openshift/api repository but lives in CCO's repo.

Reply:
> Please also re-add @JoelSpeed as API reviewer to EP. The `CredentialsRequest` should very arguably have been born into the openshift/api repository but lives in CCO's repo and in any event would need API review.

Seems done. Marking resolved